In [2]:
import numpy as np
import pandas as pd
import pickle
from yaml import safe_load
import os
from tqdm import tqdm
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import r2_score, mean_absolute_error

In [3]:
filenames = []
for file in os.listdir('data'):
    filenames.append(os.path.join('data',file))

In [4]:
filenames[0:5]

['data\\1001349.yaml',
 'data\\1001351.yaml',
 'data\\1001353.yaml',
 'data\\1004729.yaml',
 'data\\1007655.yaml']

In [5]:
final_df = pd.DataFrame()
counter = 1
dataframes = []
for file in tqdm(filenames):
    with open(file, 'r') as f:
        df = pd.json_normalize(safe_load(f))
        df['match_id'] = counter
        dataframes.append(df)  # Append the dataframe to the list
        counter += 1

# Concatenate all dataframes at once
final_df = pd.concat(dataframes, ignore_index=True)

final_df

100%|██████████| 2506/2506 [07:55<00:00,  5.27it/s]


innings  meta.data_version  \
0     [{'1st innings': {'team': 'Australia', 'delive...               0.91   
1     [{'1st innings': {'team': 'Australia', 'delive...               0.91   
2     [{'1st innings': {'team': 'Australia', 'delive...               0.91   
3     [{'1st innings': {'team': 'Hong Kong', 'delive...               0.91   
4     [{'1st innings': {'team': 'Zimbabwe', 'deliver...               0.91   
...                                                 ...                ...   
2501  [{'1st innings': {'team': 'Sri Lanka', 'delive...               0.91   
2502  [{'1st innings': {'team': 'Bangladesh', 'deliv...               0.91   
2503  [{'1st innings': {'team': 'Netherlands', 'deli...               0.91   
2504  [{'1st innings': {'team': 'Australia', 'delive...               0.91   
2505  [{'1st innings': {'team': 'Sri Lanka', 'delive...               0.91   

     meta.created  meta.revision  info.balls_per_over    info.dates  \
0      2017-02-18              2                    6  [2017-02-17]   
1      2017-02-19              2                    6  [2017-02-19]   
2      2017-02-23              1                    6  [2017-02-22]   
3      2016-09-12              1                    6  [2016-09-05]   
4      2016-06-19              1                    6  [2016-06-18]   
...           ...            ...                  ...           ...   
2501   2016-03-05              2                    6  [2016-03-04]   
2502   2016-03-08              1                    6  [2016-03-06]   
2503   2016-02-03              1                    6  [2016-02-03]   
2504   2016-09-12              1                    6  [2016-09-06]   
2505   2016-09-12              1                    6  [2016-09-09]   

     info.gender info.match_type  info.outcome.by.wickets info.outcome.winner  \
0           male             T20                      5.0           Sri Lanka   
1           male             T20                      2.0           Sri Lanka   
2           male             T20                      NaN           Australia   
3           male             T20                      NaN           Hong Kong   
4           male             T20                      NaN            Zimbabwe   
...          ...             ...                      ...                 ...   
2501        male             T20                      6.0            Pakistan   
2502        male             T20                      8.0               India   
2503        male             T20                      NaN         Netherlands   
2504        male             T20                      NaN           Australia   
2505        male             T20                      4.0           Australia   

      ...  info.registry.people.Usman Mushtaq  \
0     ...                                 NaN   
1     ...                                 NaN   
2     ...                                 NaN   
3     ...                                 NaN   
4     ...                                 NaN   
...   ...                                 ...   
2501  ...                                 NaN   
2502  ...                                 NaN   
2503  ...                                 NaN   
2504  ...                                 NaN   
2505  ...                                 NaN   

     info.registry.people.Mohammad Shahid info.registry.people.Muktar Ali  \
0                                     NaN                             NaN   
1                                     NaN                             NaN   
2                                     NaN                             NaN   
3                                     NaN                             NaN   
4                                     NaN                             NaN   
...                                   ...                             ...   
2501                                  NaN                             NaN   
2502                                  NaN                             NaN   
250

In [6]:
backup = final_df.copy()
final_df

innings  meta.data_version  \
0     [{'1st innings': {'team': 'Australia', 'delive...               0.91   
1     [{'1st innings': {'team': 'Australia', 'delive...               0.91   
2     [{'1st innings': {'team': 'Australia', 'delive...               0.91   
3     [{'1st innings': {'team': 'Hong Kong', 'delive...               0.91   
4     [{'1st innings': {'team': 'Zimbabwe', 'deliver...               0.91   
...                                                 ...                ...   
2501  [{'1st innings': {'team': 'Sri Lanka', 'delive...               0.91   
2502  [{'1st innings': {'team': 'Bangladesh', 'deliv...               0.91   
2503  [{'1st innings': {'team': 'Netherlands', 'deli...               0.91   
2504  [{'1st innings': {'team': 'Australia', 'delive...               0.91   
2505  [{'1st innings': {'team': 'Sri Lanka', 'delive...               0.91   

     meta.created  meta.revision  info.balls_per_over    info.dates  \
0      2017-02-18              2                    6  [2017-02-17]   
1      2017-02-19              2                    6  [2017-02-19]   
2      2017-02-23              1                    6  [2017-02-22]   
3      2016-09-12              1                    6  [2016-09-05]   
4      2016-06-19              1                    6  [2016-06-18]   
...           ...            ...                  ...           ...   
2501   2016-03-05              2                    6  [2016-03-04]   
2502   2016-03-08              1                    6  [2016-03-06]   
2503   2016-02-03              1                    6  [2016-02-03]   
2504   2016-09-12              1                    6  [2016-09-06]   
2505   2016-09-12              1                    6  [2016-09-09]   

     info.gender info.match_type  info.outcome.by.wickets info.outcome.winner  \
0           male             T20                      5.0           Sri Lanka   
1           male             T20                      2.0           Sri Lanka   
2           male             T20                      NaN           Australia   
3           male             T20                      NaN           Hong Kong   
4           male             T20                      NaN            Zimbabwe   
...          ...             ...                      ...                 ...   
2501        male             T20                      6.0            Pakistan   
2502        male             T20                      8.0               India   
2503        male             T20                      NaN         Netherlands   
2504        male             T20                      NaN           Australia   
2505        male             T20                      4.0           Australia   

      ...  info.registry.people.Usman Mushtaq  \
0     ...                                 NaN   
1     ...                                 NaN   
2     ...                                 NaN   
3     ...                                 NaN   
4     ...                                 NaN   
...   ...                                 ...   
2501  ...                                 NaN   
2502  ...                                 NaN   
2503  ...                                 NaN   
2504  ...                                 NaN   
2505  ...                                 NaN   

     info.registry.people.Mohammad Shahid info.registry.people.Muktar Ali  \
0                                     NaN                             NaN   
1                                     NaN                             NaN   
2                                     NaN                             NaN   
3                                     NaN                             NaN   
4                                     NaN                             NaN   
...                                   ...                             ...   
2501                                  NaN                             NaN   
2502                                  NaN                             NaN   
250

In [14]:
final_df.drop(columns=[
    'meta.data_version',
    'meta.created',
    'meta.revision',
    'info.balls_per_over',
    'info.outcome.bowl_out',
    'info.bowl_out',
    'info.supersubs.South Africa',
    'info.supersubs.New Zealand',
    'info.outcome.eliminator',
    'info.outcome.result',
    'info.outcome.method',
    'info.neutral_venue',
    'info.match_type_number',
    'info.outcome.by.runs',
    'info.outcome.by.wickets',
    'info.toss.uncontested'
],inplace=True,errors='ignore')
final_df.drop(columns=final_df.filter(like='info.registry.people').columns, inplace=True,errors='ignore')
final_df.drop(columns=final_df.filter(like='info.players').columns, inplace=True,errors='ignore')


In [15]:
final_df

innings    info.dates  \
0     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
1     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
2     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
3     [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
4     [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
...                                                 ...           ...   
2501  [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
2502  [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
2503  [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
2504  [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
2505  [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

     info.gender info.match_type info.outcome.winner  info.overs  \
0           male             T20           Sri Lanka          20   
1           male             T20           Sri Lanka          20   
2           male             T20           Australia          20   
3           male             T20           Hong Kong          20   
4           male             T20            Zimbabwe          20   
...          ...             ...                 ...         ...   
2501        male             T20            Pakistan          20   
2502        male             T20               India          20   
2503        male             T20         Netherlands          20   
2504        male             T20           Australia          20   
2505        male             T20           Australia          20   

     info.player_of_match                           info.teams  \
0         [DAS Gunaratne]               [Australia, Sri Lanka]   
1         [DAS Gunaratne]               [Australia, Sri Lanka]   
2               [A Zampa]               [Australia, Sri Lanka]   
3                     NaN                 [Ireland, Hong Kong]   
4          [E Chigumbura]                    [Zimbabwe, India]   
...                   ...                                  ...   
2501         [Umar Akmal]                [Pakistan, Sri Lanka]   
2502           [S Dhawan]                  [Bangladesh, India]   
2503   [Mudassar Bukhari]  [United Arab Emirates, Netherlands]   
2504         [GJ Maxwell]               [Sri Lanka, Australia]   
2505         [GJ Maxwell]               [Sri Lanka, Australia]   

     info.toss.decision      info.toss.winner  \
0                 field             Sri Lanka   
1                 field             Sri Lanka   
2                 field             Sri Lanka   
3                   bat             Hong Kong   
4                 field                 India   
...                 ...                   ...   
2501              field              Pakistan   
2502              field                 India   
2503              field  United Arab Emirates   
2504              field             Sri Lanka   
2505                bat             Sri Lanka   

                            info.umpires  \
0                 [MD Martell, P Wilson]   
1                  [SD Fry, SJ Nogajski]   
2                 [MD Martell, P Wilson]   
3                    [R Black, AJ Neill]   
4               [TJ Matibiri, RB Tiffin]   
...                                  ...   
2501        [AK Chaudhary, Enamul Haque]   
2502    [RSA Palliyaguruge, Shozab Raza]   
2503          [CK Nandan, Sarika Prasad]   
2504      [REJ Martinesz, RR Wimalasiri]   
2505  [REJ Martinesz, RSA Palliyaguruge]   

                                   info.venue  match_id    info.city  
0                    Melbourne Cricket Ground         1          NaN  
1              Simonds Stadium, South Geelong         2     Victoria  
2                               Adelaide Oval         3          NaN  
3           Bready Cricket Club, Magheramason         4  Londonderry  
4                          Harare Sports Club         5          NaN  
...              

In [16]:
final_df['info.gender'].value_counts()

info.gender
male    2506
Name: count, dtype: int64

In [17]:
# final_df = final_df[final_df['info.gender'] == 'male']
final_df.drop(columns=['info.gender'],inplace=True)
final_df

innings    info.dates  \
0     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
1     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
2     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
3     [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
4     [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
...                                                 ...           ...   
2501  [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
2502  [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
2503  [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
2504  [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
2505  [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

     info.match_type info.outcome.winner  info.overs info.player_of_match  \
0                T20           Sri Lanka          20      [DAS Gunaratne]   
1                T20           Sri Lanka          20      [DAS Gunaratne]   
2                T20           Australia          20            [A Zampa]   
3                T20           Hong Kong          20                  NaN   
4                T20            Zimbabwe          20       [E Chigumbura]   
...              ...                 ...         ...                  ...   
2501             T20            Pakistan          20         [Umar Akmal]   
2502             T20               India          20           [S Dhawan]   
2503             T20         Netherlands          20   [Mudassar Bukhari]   
2504             T20           Australia          20         [GJ Maxwell]   
2505             T20           Australia          20         [GJ Maxwell]   

                               info.teams info.toss.decision  \
0                  [Australia, Sri Lanka]              field   
1                  [Australia, Sri Lanka]              field   
2                  [Australia, Sri Lanka]              field   
3                    [Ireland, Hong Kong]                bat   
4                       [Zimbabwe, India]              field   
...                                   ...                ...   
2501                [Pakistan, Sri Lanka]              field   
2502                  [Bangladesh, India]              field   
2503  [United Arab Emirates, Netherlands]              field   
2504               [Sri Lanka, Australia]              field   
2505               [Sri Lanka, Australia]                bat   

          info.toss.winner                        info.umpires  \
0                Sri Lanka              [MD Martell, P Wilson]   
1                Sri Lanka               [SD Fry, SJ Nogajski]   
2                Sri Lanka              [MD Martell, P Wilson]   
3                Hong Kong                 [R Black, AJ Neill]   
4                    India            [TJ Matibiri, RB Tiffin]   
...                    ...                                 ...   
2501              Pakistan        [AK Chaudhary, Enamul Haque]   
2502                 India    [RSA Palliyaguruge, Shozab Raza]   
2503  United Arab Emirates          [CK Nandan, Sarika Prasad]   
2504             Sri Lanka      [REJ Martinesz, RR Wimalasiri]   
2505             Sri Lanka  [REJ Martinesz, RSA Palliyaguruge]   

                                   info.venue  match_id    info.city  
0                    Melbourne Cricket Ground         1          NaN  
1              Simonds Stadium, South Geelong         2     Victoria  
2                               Adelaide Oval         3          NaN  
3           Bready Cricket Club, Magheramason         4  Londonderry  
4                          Harare Sports Club         5          NaN  
...                                       ...       ...          ...  
2501            Shere Bangla National Stadium      2502       Mirpur  
2502            Shere Bangla National Stadium      2503       Mirpur  
2503                              ICC Academy

In [18]:
final_df['info.match_type'].value_counts()

info.match_type
T20    2506
Name: count, dtype: int64

In [19]:
final_df['info.overs'].value_counts()

info.overs
20    2503
50       3
Name: count, dtype: int64

In [20]:
final_df = final_df[final_df['info.overs'] == 20]
final_df.drop(columns=['info.overs','info.match_type'],inplace=True)
final_df

C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_6280\3157072726.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df.drop(columns=['info.overs','info.match_type'],inplace=True)


innings    info.dates  \
0     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-17]   
1     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-19]   
2     [{'1st innings': {'team': 'Australia', 'delive...  [2017-02-22]   
3     [{'1st innings': {'team': 'Hong Kong', 'delive...  [2016-09-05]   
4     [{'1st innings': {'team': 'Zimbabwe', 'deliver...  [2016-06-18]   
...                                                 ...           ...   
2501  [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-03-04]   
2502  [{'1st innings': {'team': 'Bangladesh', 'deliv...  [2016-03-06]   
2503  [{'1st innings': {'team': 'Netherlands', 'deli...  [2016-02-03]   
2504  [{'1st innings': {'team': 'Australia', 'delive...  [2016-09-06]   
2505  [{'1st innings': {'team': 'Sri Lanka', 'delive...  [2016-09-09]   

     info.outcome.winner info.player_of_match  \
0              Sri Lanka      [DAS Gunaratne]   
1              Sri Lanka      [DAS Gunaratne]   
2              Australia            [A Zampa]   
3              Hong Kong                  NaN   
4               Zimbabwe       [E Chigumbura]   
...                  ...                  ...   
2501            Pakistan         [Umar Akmal]   
2502               India           [S Dhawan]   
2503         Netherlands   [Mudassar Bukhari]   
2504           Australia         [GJ Maxwell]   
2505           Australia         [GJ Maxwell]   

                               info.teams info.toss.decision  \
0                  [Australia, Sri Lanka]              field   
1                  [Australia, Sri Lanka]              field   
2                  [Australia, Sri Lanka]              field   
3                    [Ireland, Hong Kong]                bat   
4                       [Zimbabwe, India]              field   
...                                   ...                ...   
2501                [Pakistan, Sri Lanka]              field   
2502                  [Bangladesh, India]              field   
2503  [United Arab Emirates, Netherlands]              field   
2504               [Sri Lanka, Australia]              field   
2505               [Sri Lanka, Australia]                bat   

          info.toss.winner                        info.umpires  \
0                Sri Lanka              [MD Martell, P Wilson]   
1                Sri Lanka               [SD Fry, SJ Nogajski]   
2                Sri Lanka              [MD Martell, P Wilson]   
3                Hong Kong                 [R Black, AJ Neill]   
4                    India            [TJ Matibiri, RB Tiffin]   
...                    ...                                 ...   
2501              Pakistan        [AK Chaudhary, Enamul Haque]   
2502                 India    [RSA Palliyaguruge, Shozab Raza]   
2503  United Arab Emirates          [CK Nandan, Sarika Prasad]   
2504             Sri Lanka      [REJ Martinesz, RR Wimalasiri]   
2505             Sri Lanka  [REJ Martinesz, RSA Palliyaguruge]   

                                   info.venue  match_id    info.city  
0                    Melbourne Cricket Ground         1          NaN  
1              Simonds Stadium, South Geelong         2     Victoria  
2                               Adelaide Oval         3          NaN  
3           Bready Cricket Club, Magheramason         4  Londonderry  
4                          Harare Sports Club         5          NaN  
...                                       ...       ...          ...  
2501            Shere Bangla National Stadium      2502       Mirpur  
2502            Shere Bangla National Stadium      2503       Mirpur  
2503                              ICC Academy      2504        Dubai  
2504  Pallekele International Cricket Stadium      2505          NaN  
2505                      R Premadasa Stadium      2506      Colombo  

[2503 rows x 11 columns]

In [21]:
pickle.dump(final_df,open('dataset_level1.pkl','wb'))

In [35]:
matches = pickle.load(open('dataset_level1.pkl','rb'))

matches.iloc[0]['innings'][0]['1st innings']['deliveries']

[{0.1: {'batsman': 'AJ Finch',
   'bowler': 'SL Malinga',
   'non_striker': 'M Klinger',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.2: {'batsman': 'AJ Finch',
   'bowler': 'SL Malinga',
   'non_striker': 'M Klinger',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.3: {'batsman': 'AJ Finch',
   'bowler': 'SL Malinga',
   'non_striker': 'M Klinger',
   'runs': {'batsman': 1, 'extras': 0, 'total': 1}}},
 {0.4: {'batsman': 'M Klinger',
   'bowler': 'SL Malinga',
   'non_striker': 'AJ Finch',
   'runs': {'batsman': 2, 'extras': 0, 'total': 2}}},
 {0.5: {'batsman': 'M Klinger',
   'bowler': 'SL Malinga',
   'non_striker': 'AJ Finch',
   'runs': {'batsman': 0, 'extras': 0, 'total': 0}}},
 {0.6: {'batsman': 'M Klinger',
   'bowler': 'SL Malinga',
   'non_striker': 'AJ Finch',
   'runs': {'batsman': 3, 'extras': 0, 'total': 3}}},
 {1.1: {'batsman': 'M Klinger',
   'bowler': 'KMDN Kulasekara',
   'non_striker': 'AJ Finch',
   'runs': {'batsman': 0, 'extras': 0, 'total': 

In [36]:
count = 1
delivery_df = pd.DataFrame()  # This will hold the final data
dataframes = []  # A list to store dataframes for concatenation later

for index, row in matches.iterrows():
    if count in [75, 108, 150, 180, 268, 360, 443, 458, 584, 748, 982, 1052, 1111, 1226, 1345]:
        count += 1
        continue
    count += 1
    
    # Temporary lists to hold data for each ball
    ball_of_match = []
    batsman = []
    bowler = []
    runs = []
    player_of_dismissed = []
    teams = []
    batting_team = []
    match_id = []
    city = []
    venue = []
    
    for ball in row['innings'][0]['1st innings']['deliveries']:
        for key in ball.keys():
            match_id.append(count)
            batting_team.append(row['innings'][0]['1st innings']['team'])
            teams.append(row['info.teams'])
            ball_of_match.append(key)
            batsman.append(ball[key]['batsman'])
            bowler.append(ball[key]['bowler'])
            runs.append(ball[key]['runs']['total'])
            city.append(row['info.city'])
            venue.append(row['info.venue'])
            try:
                player_of_dismissed.append(ball[key]['wicket']['player_out'])
            except KeyError:
                player_of_dismissed.append('0')
    
    # Create a dataframe for this loop's data
    loop_df = pd.DataFrame({
        'match_id': match_id,
        'teams': teams,
        'batting_team': batting_team,
        'ball': ball_of_match,
        'batsman': batsman,
        'bowler': bowler,
        'runs': runs,
        'player_dismissed': player_of_dismissed,
        'city': city,
        'venue': venue
    })
    
    # Append this loop's dataframe to the list of dataframes
    dataframes.append(loop_df)

# Concatenate all dataframes once the loop is done
delivery_df = pd.concat(dataframes, ignore_index=True)

delivery_df


match_id                   teams batting_team  ball          batsman  \
0              2  [Australia, Sri Lanka]    Australia   0.1         AJ Finch   
1              2  [Australia, Sri Lanka]    Australia   0.2         AJ Finch   
2              2  [Australia, Sri Lanka]    Australia   0.3         AJ Finch   
3              2  [Australia, Sri Lanka]    Australia   0.4        M Klinger   
4              2  [Australia, Sri Lanka]    Australia   0.5        M Klinger   
...          ...                     ...          ...   ...              ...   
301349      2504  [Sri Lanka, Australia]    Sri Lanka  19.3  SMSM Senanayake   
301350      2504  [Sri Lanka, Australia]    Sri Lanka  19.4      DM de Silva   
301351      2504  [Sri Lanka, Australia]    Sri Lanka  19.5      DM de Silva   
301352      2504  [Sri Lanka, Australia]    Sri Lanka  19.6  SMSM Senanayake   
301353      2504  [Sri Lanka, Australia]    Sri Lanka  19.7  SMSM Senanayake   

            bowler  runs player_dismissed     city                     venue  
0       SL Malinga     0                0      NaN  Melbourne Cricket Ground  
1       SL Malinga     0                0      NaN  Melbourne Cricket Ground  
2       SL Malinga     1                0      NaN  Melbourne Cricket Ground  
3       SL Malinga     2                0      NaN  Melbourne Cricket Ground  
4       SL Malinga     0                0      NaN  Melbourne Cricket Ground  
...            ...   ...              ...      ...                       ...  
301349    MA Starc     1                0  Colombo       R Premadasa Stadium  
301350    MA Starc     0                0  Colombo       R Premadasa Stadium  
301351    MA Starc     0      DM de Silva  Colombo       R Premadasa Stadium  
301352    MA Starc     2                0  Colombo       R Premadasa Stadium  
301353    MA Starc     1                0  Colombo       R Premadasa Stadium  

[301354 rows x 10 columns]

In [37]:
def bowl(row):
    for team in row['teams']:
        if team != row['batting_team']:
            return team

In [38]:
delivery_df['bowling_team'] = delivery_df.apply(bowl,axis=1)

In [39]:
delivery_df

match_id                   teams batting_team  ball          batsman  \
0              2  [Australia, Sri Lanka]    Australia   0.1         AJ Finch   
1              2  [Australia, Sri Lanka]    Australia   0.2         AJ Finch   
2              2  [Australia, Sri Lanka]    Australia   0.3         AJ Finch   
3              2  [Australia, Sri Lanka]    Australia   0.4        M Klinger   
4              2  [Australia, Sri Lanka]    Australia   0.5        M Klinger   
...          ...                     ...          ...   ...              ...   
301349      2504  [Sri Lanka, Australia]    Sri Lanka  19.3  SMSM Senanayake   
301350      2504  [Sri Lanka, Australia]    Sri Lanka  19.4      DM de Silva   
301351      2504  [Sri Lanka, Australia]    Sri Lanka  19.5      DM de Silva   
301352      2504  [Sri Lanka, Australia]    Sri Lanka  19.6  SMSM Senanayake   
301353      2504  [Sri Lanka, Australia]    Sri Lanka  19.7  SMSM Senanayake   

            bowler  runs player_dismissed     city                     venue  \
0       SL Malinga     0                0      NaN  Melbourne Cricket Ground   
1       SL Malinga     0                0      NaN  Melbourne Cricket Ground   
2       SL Malinga     1                0      NaN  Melbourne Cricket Ground   
3       SL Malinga     2                0      NaN  Melbourne Cricket Ground   
4       SL Malinga     0                0      NaN  Melbourne Cricket Ground   
...            ...   ...              ...      ...                       ...   
301349    MA Starc     1                0  Colombo       R Premadasa Stadium   
301350    MA Starc     0                0  Colombo       R Premadasa Stadium   
301351    MA Starc     0      DM de Silva  Colombo       R Premadasa Stadium   
301352    MA Starc     2                0  Colombo       R Premadasa Stadium   
301353    MA Starc     1                0  Colombo       R Premadasa Stadium   

       bowling_team  
0         Sri Lanka  
1         Sri Lanka  
2         Sri Lanka  
3         Sri Lanka  
4         Sri Lanka  
...             ...  
301349    Australia  
301350    Australia  
301351    Australia  
301352    Australia  
301353    Australia  

[301354 rows x 11 columns]

In [40]:
delivery_df.drop(columns=['teams'],inplace=True)

In [41]:
delivery_df['batting_team'].unique()

array(['Australia', 'Hong Kong', 'Zimbabwe', 'India', 'Bangladesh',
       'New Zealand', 'South Africa', 'England', 'West Indies', 'Ireland',
       'Afghanistan', 'Pakistan', 'United Arab Emirates', 'Scotland',
       'Oman', 'Papua New Guinea', 'Sri Lanka', 'Netherlands', 'Nepal',
       'Vanuatu', 'Philippines', 'United States of America', 'Germany',
       'Ghana', 'Uganda', 'Kenya', 'Namibia', 'Nigeria', 'Botswana',
       'Guernsey', 'Denmark', 'Jersey', 'Italy', 'Norway', 'Thailand',
       'Malaysia', 'Maldives', 'Singapore', 'Kuwait', 'Bermuda', 'Canada',
       'Cayman Islands', 'Portugal', 'Gibraltar', 'Spain', 'Bhutan',
       'Qatar', 'Iran', 'Belgium', 'Isle of Man', 'Bulgaria', 'Romania',
       'Luxembourg', 'Austria', 'Czech Republic', 'Greece', 'Serbia',
       'Malta', 'France', 'Sweden', 'Rwanda', 'Finland', 'Hungary',
       'Estonia', 'Cyprus', 'Switzerland', 'Seychelles', 'Malawi',
       'Lesotho', 'Swaziland', 'Tanzania', 'Mozambique', 'Sierra Leone',
       '

In [42]:
teams = [
    'Australia',
    'India',
    'Bangladesh',
    'New Zealand',
    'South Africa',
    'England',
    'West Indies',
    'Afghanistan',
    'Pakistan',
    'Sri Lanka'    
]

In [43]:
delivery_df = delivery_df[delivery_df['batting_team'].isin(teams)]
delivery_df = delivery_df[delivery_df['bowling_team'].isin(teams)]

In [44]:
delivery_df

match_id batting_team  ball          batsman      bowler  runs  \
0              2    Australia   0.1         AJ Finch  SL Malinga     0   
1              2    Australia   0.2         AJ Finch  SL Malinga     0   
2              2    Australia   0.3         AJ Finch  SL Malinga     1   
3              2    Australia   0.4        M Klinger  SL Malinga     2   
4              2    Australia   0.5        M Klinger  SL Malinga     0   
...          ...          ...   ...              ...         ...   ...   
301349      2504    Sri Lanka  19.3  SMSM Senanayake    MA Starc     1   
301350      2504    Sri Lanka  19.4      DM de Silva    MA Starc     0   
301351      2504    Sri Lanka  19.5      DM de Silva    MA Starc     0   
301352      2504    Sri Lanka  19.6  SMSM Senanayake    MA Starc     2   
301353      2504    Sri Lanka  19.7  SMSM Senanayake    MA Starc     1   

       player_dismissed     city                     venue bowling_team  
0                     0      NaN  Melbourne Cricket Ground    Sri Lanka  
1                     0      NaN  Melbourne Cricket Ground    Sri Lanka  
2                     0      NaN  Melbourne Cricket Ground    Sri Lanka  
3                     0      NaN  Melbourne Cricket Ground    Sri Lanka  
4                     0      NaN  Melbourne Cricket Ground    Sri Lanka  
...                 ...      ...                       ...          ...  
301349                0  Colombo       R Premadasa Stadium    Australia  
301350                0  Colombo       R Premadasa Stadium    Australia  
301351      DM de Silva  Colombo       R Premadasa Stadium    Australia  
301352                0  Colombo       R Premadasa Stadium    Australia  
301353                0  Colombo       R Premadasa Stadium    Australia  

[100582 rows x 10 columns]

In [45]:
# delivery_df['venue'].unique()

array(['Melbourne Cricket Ground', 'Simonds Stadium, South Geelong',
       'Adelaide Oval', 'McLean Park', 'Bay Oval', 'Eden Park',
       'The Rose Bowl', 'County Ground', 'Sophia Gardens',
       'Riverside Ground', 'Green Park',
       'Vidarbha Cricket Association Stadium, Jamtha',
       'M Chinnaswamy Stadium',
       'Central Broward Regional Park Stadium Turf Ground',
       'Dubai International Cricket Stadium', 'Sheikh Zayed Stadium',
       'Sydney Cricket Ground', 'Bellerive Oval', 'Westpac Stadium',
       'Seddon Park', 'Mangaung Oval', 'Senwes Park',
       'Kensington Oval, Bridgetown', "Queen's Park Oval, Port of Spain",
       'R Premadasa Stadium', 'Warner Park, Basseterre',
       'Sabina Park, Kingston', 'R.Premadasa Stadium, Khettarama',
       'Saxton Oval', 'JSCA International Stadium Complex', 'Edgbaston',
       'Old Trafford', 'Arun Jaitley Stadium',
       'Saurashtra Cricket Association Stadium',
       'Greenfield International Stadium', 'Gaddafi Stadium'

In [33]:
output = delivery_df[['match_id','batting_team','bowling_team','ball','runs','player_dismissed','city','venue']]

In [34]:
output

match_id batting_team bowling_team  ball  runs player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                0   
1              2    Australia    Sri Lanka   0.2     0                0   
2              2    Australia    Sri Lanka   0.3     1                0   
3              2    Australia    Sri Lanka   0.4     2                0   
4              2    Australia    Sri Lanka   0.5     0                0   
...          ...          ...          ...   ...   ...              ...   
301349      2504    Sri Lanka    Australia  19.3     1                0   
301350      2504    Sri Lanka    Australia  19.4     0                0   
301351      2504    Sri Lanka    Australia  19.5     0      DM de Silva   
301352      2504    Sri Lanka    Australia  19.6     2                0   
301353      2504    Sri Lanka    Australia  19.7     1                0   

           city                     venue  
0           NaN  Melbourne Cricket Ground  
1           NaN  Melbourne Cricket Ground  
2           NaN  Melbourne Cricket Ground  
3           NaN  Melbourne Cricket Ground  
4           NaN  Melbourne Cricket Ground  
...         ...                       ...  
301349  Colombo       R Premadasa Stadium  
301350  Colombo       R Premadasa Stadium  
301351  Colombo       R Premadasa Stadium  
301352  Colombo       R Premadasa Stadium  
301353  Colombo       R Premadasa Stadium  

[100582 rows x 8 columns]

In [35]:
pickle.dump(output,open('dataset_level2.pkl','wb'))

In [3]:
df = pickle.load(open('dataset_level2.pkl','rb'))

In [4]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                0   
1              2    Australia    Sri Lanka   0.2     0                0   
2              2    Australia    Sri Lanka   0.3     1                0   
3              2    Australia    Sri Lanka   0.4     2                0   
4              2    Australia    Sri Lanka   0.5     0                0   
...          ...          ...          ...   ...   ...              ...   
301349      2504    Sri Lanka    Australia  19.3     1                0   
301350      2504    Sri Lanka    Australia  19.4     0                0   
301351      2504    Sri Lanka    Australia  19.5     0      DM de Silva   
301352      2504    Sri Lanka    Australia  19.6     2                0   
301353      2504    Sri Lanka    Australia  19.7     1                0   

           city                     venue  
0           NaN  Melbourne Cricket Ground  
1           NaN  Melbourne Cricket Ground  
2           NaN  Melbourne Cricket Ground  
3           NaN  Melbourne Cricket Ground  
4           NaN  Melbourne Cricket Ground  
...         ...                       ...  
301349  Colombo       R Premadasa Stadium  
301350  Colombo       R Premadasa Stadium  
301351  Colombo       R Premadasa Stadium  
301352  Colombo       R Premadasa Stadium  
301353  Colombo       R Premadasa Stadium  

[100582 rows x 8 columns]

In [5]:
df.isnull().sum()


match_id               0
batting_team           0
bowling_team           0
ball                   0
runs                   0
player_dismissed       0
city                8671
venue                  0
dtype: int64

In [6]:
df[df['city'].isnull()]['venue'].value_counts()


venue
Dubai International Cricket Stadium        3092
Pallekele International Cricket Stadium    2066
Melbourne Cricket Ground                   1453
Sydney Cricket Ground                       749
Adelaide Oval                               498
Harare Sports Club                          372
Sharjah Cricket Stadium                     249
Sylhet International Cricket Stadium        128
Carrara Oval                                 64
Name: count, dtype: int64

In [7]:
cities = np.where(df['city'].isnull(),df['venue'].str.split().apply(lambda x:x[0]),df['city'])

In [8]:
df['city'] = cities

In [9]:
df.isnull().sum()

match_id            0
batting_team        0
bowling_team        0
ball                0
runs                0
player_dismissed    0
city                0
venue               0
dtype: int64

In [10]:
df.drop(columns=['venue'],inplace=True)

In [11]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                0   
1              2    Australia    Sri Lanka   0.2     0                0   
2              2    Australia    Sri Lanka   0.3     1                0   
3              2    Australia    Sri Lanka   0.4     2                0   
4              2    Australia    Sri Lanka   0.5     0                0   
...          ...          ...          ...   ...   ...              ...   
301349      2504    Sri Lanka    Australia  19.3     1                0   
301350      2504    Sri Lanka    Australia  19.4     0                0   
301351      2504    Sri Lanka    Australia  19.5     0      DM de Silva   
301352      2504    Sri Lanka    Australia  19.6     2                0   
301353      2504    Sri Lanka    Australia  19.7     1                0   

             city  
0       Melbourne  
1       Melbourne  
2       Melbourne  
3       Melbourne  
4       Melbourne  
...           ...  
301349    Colombo  
301350    Colombo  
301351    Colombo  
301352    Colombo  
301353    Colombo  

[100582 rows x 7 columns]

In [12]:
eligible_cities = df['city'].value_counts()[df['city'].value_counts() > 600].index.tolist()
df = df[df['city'].isin(eligible_cities)]
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                0   
1              2    Australia    Sri Lanka   0.2     0                0   
2              2    Australia    Sri Lanka   0.3     1                0   
3              2    Australia    Sri Lanka   0.4     2                0   
4              2    Australia    Sri Lanka   0.5     0                0   
...          ...          ...          ...   ...   ...              ...   
301349      2504    Sri Lanka    Australia  19.3     1                0   
301350      2504    Sri Lanka    Australia  19.4     0                0   
301351      2504    Sri Lanka    Australia  19.5     0      DM de Silva   
301352      2504    Sri Lanka    Australia  19.6     2                0   
301353      2504    Sri Lanka    Australia  19.7     1                0   

             city  
0       Melbourne  
1       Melbourne  
2       Melbourne  
3       Melbourne  
4       Melbourne  
...           ...  
301349    Colombo  
301350    Colombo  
301351    Colombo  
301352    Colombo  
301353    Colombo  

[84831 rows x 7 columns]

In [13]:
# Ensure that the 'runs' column is numeric
df['runs'] = pd.to_numeric(df['runs'], errors='coerce')

# Fill any NaN values with 0 (optional, depending on your dataset)
df['runs'] = df['runs'].fillna(0)

# Now, apply cumsum() within each group (match_id)
df['current_score'] = df.groupby('match_id')['runs'].cumsum()

# Check the result
print(df[['match_id', 'runs', 'current_score']].head())


   match_id  runs  current_score
0         2     0              0
1         2     0              0
2         2     1              1
3         2     2              3
4         2     0              3


C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\2417348336.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['runs'] = pd.to_numeric(df['runs'], errors='coerce')
C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\2417348336.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['runs'] = df['runs'].fillna(0)
C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\2417348336.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,c

In [14]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                0   
1              2    Australia    Sri Lanka   0.2     0                0   
2              2    Australia    Sri Lanka   0.3     1                0   
3              2    Australia    Sri Lanka   0.4     2                0   
4              2    Australia    Sri Lanka   0.5     0                0   
...          ...          ...          ...   ...   ...              ...   
301349      2504    Sri Lanka    Australia  19.3     1                0   
301350      2504    Sri Lanka    Australia  19.4     0                0   
301351      2504    Sri Lanka    Australia  19.5     0      DM de Silva   
301352      2504    Sri Lanka    Australia  19.6     2                0   
301353      2504    Sri Lanka    Australia  19.7     1                0   

             city  current_score  
0       Melbourne              0  
1       Melbourne              0  
2       Melbourne              1  
3       Melbourne              3  
4       Melbourne              3  
...           ...            ...  
301349    Colombo            125  
301350    Colombo            125  
301351    Colombo            125  
301352    Colombo            127  
301353    Colombo            128  

[84831 rows x 8 columns]

In [15]:
df['over'] = df['ball'].apply(lambda x:str(x).split(".")[0])
df['ball_no'] = df['ball'].apply(lambda x:str(x).split(".")[1])
df

C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\2777558413.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['over'] = df['ball'].apply(lambda x:str(x).split(".")[0])
C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\2777558413.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['ball_no'] = df['ball'].apply(lambda x:str(x).split(".")[1])


match_id batting_team bowling_team  ball  runs player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                0   
1              2    Australia    Sri Lanka   0.2     0                0   
2              2    Australia    Sri Lanka   0.3     1                0   
3              2    Australia    Sri Lanka   0.4     2                0   
4              2    Australia    Sri Lanka   0.5     0                0   
...          ...          ...          ...   ...   ...              ...   
301349      2504    Sri Lanka    Australia  19.3     1                0   
301350      2504    Sri Lanka    Australia  19.4     0                0   
301351      2504    Sri Lanka    Australia  19.5     0      DM de Silva   
301352      2504    Sri Lanka    Australia  19.6     2                0   
301353      2504    Sri Lanka    Australia  19.7     1                0   

             city  current_score over ball_no  
0       Melbourne              0    0       1  
1       Melbourne              0    0       2  
2       Melbourne              1    0       3  
3       Melbourne              3    0       4  
4       Melbourne              3    0       5  
...           ...            ...  ...     ...  
301349    Colombo            125   19       3  
301350    Colombo            125   19       4  
301351    Colombo            125   19       5  
301352    Colombo            127   19       6  
301353    Colombo            128   19       7  

[84831 rows x 10 columns]

In [16]:
df['balls_bowled'] = (df['over'].astype('int')*6) + df['ball_no'].astype('int')
df

C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\562775579.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_bowled'] = (df['over'].astype('int')*6) + df['ball_no'].astype('int')


match_id batting_team bowling_team  ball  runs player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                0   
1              2    Australia    Sri Lanka   0.2     0                0   
2              2    Australia    Sri Lanka   0.3     1                0   
3              2    Australia    Sri Lanka   0.4     2                0   
4              2    Australia    Sri Lanka   0.5     0                0   
...          ...          ...          ...   ...   ...              ...   
301349      2504    Sri Lanka    Australia  19.3     1                0   
301350      2504    Sri Lanka    Australia  19.4     0                0   
301351      2504    Sri Lanka    Australia  19.5     0      DM de Silva   
301352      2504    Sri Lanka    Australia  19.6     2                0   
301353      2504    Sri Lanka    Australia  19.7     1                0   

             city  current_score over ball_no  balls_bowled  
0       Melbourne              0    0       1             1  
1       Melbourne              0    0       2             2  
2       Melbourne              1    0       3             3  
3       Melbourne              3    0       4             4  
4       Melbourne              3    0       5             5  
...           ...            ...  ...     ...           ...  
301349    Colombo            125   19       3           117  
301350    Colombo            125   19       4           118  
301351    Colombo            125   19       5           119  
301352    Colombo            127   19       6           120  
301353    Colombo            128   19       7           121  

[84831 rows x 11 columns]

In [17]:
df['balls_left'] = 120 - df['balls_bowled']
df['balls_left'] = df['balls_left'].apply(lambda x:0 if x<0 else x)
df

C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\2304996896.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_left'] = 120 - df['balls_bowled']
C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\2304996896.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_left'] = df['balls_left'].apply(lambda x:0 if x<0 else x)


match_id batting_team bowling_team  ball  runs player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                0   
1              2    Australia    Sri Lanka   0.2     0                0   
2              2    Australia    Sri Lanka   0.3     1                0   
3              2    Australia    Sri Lanka   0.4     2                0   
4              2    Australia    Sri Lanka   0.5     0                0   
...          ...          ...          ...   ...   ...              ...   
301349      2504    Sri Lanka    Australia  19.3     1                0   
301350      2504    Sri Lanka    Australia  19.4     0                0   
301351      2504    Sri Lanka    Australia  19.5     0      DM de Silva   
301352      2504    Sri Lanka    Australia  19.6     2                0   
301353      2504    Sri Lanka    Australia  19.7     1                0   

             city  current_score over ball_no  balls_bowled  balls_left  
0       Melbourne              0    0       1             1         119  
1       Melbourne              0    0       2             2         118  
2       Melbourne              1    0       3             3         117  
3       Melbourne              3    0       4             4         116  
4       Melbourne              3    0       5             5         115  
...           ...            ...  ...     ...           ...         ...  
301349    Colombo            125   19       3           117           3  
301350    Colombo            125   19       4           118           2  
301351    Colombo            125   19       5           119           1  
301352    Colombo            127   19       6           120           0  
301353    Colombo            128   19       7           121           0  

[84831 rows x 12 columns]

In [18]:
# Convert 'player_dismissed' column to 0 and 1
df['player_dismissed'] = df['player_dismissed'].apply(lambda x: 0 if x == '0' else 1)

# Ensure the column is of type int
df['player_dismissed'] = df['player_dismissed'].astype(int)

# Perform cumsum for each match_id
df['player_dismissed_cumsum'] = df.groupby('match_id')['player_dismissed'].cumsum()

# Calculate wickets left
df['wickets_left'] = 10 - df['player_dismissed_cumsum']

# Check the result
print(df[['match_id', 'player_dismissed', 'player_dismissed_cumsum', 'wickets_left']].head())


   match_id  player_dismissed  player_dismissed_cumsum  wickets_left
0         2                 0                        0            10
1         2                 0                        0            10
2         2                 0                        0            10
3         2                 0                        0            10
4         2                 0                        0            10


C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\3414697703.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['player_dismissed'] = df['player_dismissed'].apply(lambda x: 0 if x == '0' else 1)
C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\3414697703.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['player_dismissed'] = df['player_dismissed'].astype(int)
C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\3414697703.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy o

In [19]:
df

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                 0   
1              2    Australia    Sri Lanka   0.2     0                 0   
2              2    Australia    Sri Lanka   0.3     1                 0   
3              2    Australia    Sri Lanka   0.4     2                 0   
4              2    Australia    Sri Lanka   0.5     0                 0   
...          ...          ...          ...   ...   ...               ...   
301349      2504    Sri Lanka    Australia  19.3     1                 0   
301350      2504    Sri Lanka    Australia  19.4     0                 0   
301351      2504    Sri Lanka    Australia  19.5     0                 1   
301352      2504    Sri Lanka    Australia  19.6     2                 0   
301353      2504    Sri Lanka    Australia  19.7     1                 0   

             city  current_score over ball_no  balls_bowled  balls_left  \
0       Melbourne              0    0       1             1         119   
1       Melbourne              0    0       2             2         118   
2       Melbourne              1    0       3             3         117   
3       Melbourne              3    0       4             4         116   
4       Melbourne              3    0       5             5         115   
...           ...            ...  ...     ...           ...         ...   
301349    Colombo            125   19       3           117           3   
301350    Colombo            125   19       4           118           2   
301351    Colombo            125   19       5           119           1   
301352    Colombo            127   19       6           120           0   
301353    Colombo            128   19       7           121           0   

        player_dismissed_cumsum  wickets_left  
0                             0            10  
1                             0            10  
2                             0            10  
3                             0            10  
4                             0            10  
...                         ...           ...  
301349                        8             2  
301350                        8             2  
301351                        9             1  
301352                        9             1  
301353                        9             1  

[84831 rows x 14 columns]

In [20]:
df['crr'] = (df['current_score']*6)/df['balls_bowled']


C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\2902028393.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['crr'] = (df['current_score']*6)/df['balls_bowled']


In [21]:
df

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0              2    Australia    Sri Lanka   0.1     0                 0   
1              2    Australia    Sri Lanka   0.2     0                 0   
2              2    Australia    Sri Lanka   0.3     1                 0   
3              2    Australia    Sri Lanka   0.4     2                 0   
4              2    Australia    Sri Lanka   0.5     0                 0   
...          ...          ...          ...   ...   ...               ...   
301349      2504    Sri Lanka    Australia  19.3     1                 0   
301350      2504    Sri Lanka    Australia  19.4     0                 0   
301351      2504    Sri Lanka    Australia  19.5     0                 1   
301352      2504    Sri Lanka    Australia  19.6     2                 0   
301353      2504    Sri Lanka    Australia  19.7     1                 0   

             city  current_score over ball_no  balls_bowled  balls_left  \
0       Melbourne              0    0       1             1         119   
1       Melbourne              0    0       2             2         118   
2       Melbourne              1    0       3             3         117   
3       Melbourne              3    0       4             4         116   
4       Melbourne              3    0       5             5         115   
...           ...            ...  ...     ...           ...         ...   
301349    Colombo            125   19       3           117           3   
301350    Colombo            125   19       4           118           2   
301351    Colombo            125   19       5           119           1   
301352    Colombo            127   19       6           120           0   
301353    Colombo            128   19       7           121           0   

        player_dismissed_cumsum  wickets_left       crr  
0                             0            10  0.000000  
1                             0            10  0.000000  
2                             0            10  2.000000  
3                             0            10  4.500000  
4                             0            10  3.600000  
...                         ...           ...       ...  
301349                        8             2  6.410256  
301350                        8             2  6.355932  
301351                        9             1  6.302521  
301352                        9             1  6.350000  
301353                        9             1  6.347107  

[84831 rows x 15 columns]

In [22]:
# Convert 'runs' column to numeric, invalid parsing will be set as NaN
df['runs'] = pd.to_numeric(df['runs'], errors='coerce')

# Group the data by 'match_id'
groups = df.groupby('match_id')

# Get unique match IDs
match_ids = df['match_id'].unique()

last_five = []
for id in match_ids:
    # Get the group for the match and perform rolling sum on 'runs'
    runs_rolling_sum = groups.get_group(id)['runs'].rolling(window=30, min_periods=1).sum()
    # Extend the result to the list
    last_five.extend(runs_rolling_sum.values.tolist())

# Check the result
print(last_five)


C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\1593188796.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['runs'] = pd.to_numeric(df['runs'], errors='coerce')


[0.0, 0.0, 1.0, 3.0, 3.0, 6.0, 6.0, 7.0, 7.0, 7.0, 11.0, 13.0, 14.0, 15.0, 15.0, 15.0, 19.0, 19.0, 19.0, 20.0, 21.0, 22.0, 26.0, 27.0, 30.0, 33.0, 37.0, 41.0, 42.0, 43.0, 44.0, 45.0, 44.0, 42.0, 43.0, 41.0, 41.0, 41.0, 42.0, 42.0, 39.0, 38.0, 39.0, 38.0, 39.0, 40.0, 37.0, 41.0, 42.0, 42.0, 44.0, 44.0, 40.0, 39.0, 37.0, 35.0, 31.0, 27.0, 26.0, 31.0, 31.0, 30.0, 31.0, 31.0, 32.0, 31.0, 31.0, 36.0, 37.0, 37.0, 37.0, 37.0, 36.0, 36.0, 36.0, 36.0, 35.0, 32.0, 32.0, 31.0, 29.0, 29.0, 31.0, 32.0, 32.0, 32.0, 33.0, 36.0, 37.0, 37.0, 39.0, 41.0, 40.0, 41.0, 41.0, 42.0, 44.0, 39.0, 43.0, 45.0, 48.0, 48.0, 48.0, 49.0, 49.0, 49.0, 51.0, 51.0, 51.0, 52.0, 52.0, 53.0, 53.0, 53.0, 54.0, 53.0, 52.0, 50.0, 53.0, 47.0, 46.0, 45.0, 46.0, 47.0, 49.0, 0.0, 0.0, 0.0, 0.0, 1.0, 1.0, 2.0, 8.0, 9.0, 9.0, 13.0, 14.0, 15.0, 15.0, 19.0, 20.0, 21.0, 21.0, 25.0, 26.0, 27.0, 31.0, 31.0, 31.0, 35.0, 35.0, 39.0, 40.0, 44.0, 46.0, 52.0, 53.0, 57.0, 58.0, 59.0, 59.0, 59.0, 53.0, 53.0, 54.0, 50.0, 50.0, 49.0, 50.0, 47.0,

In [23]:
df['last_five'] = last_five


C:\Users\NIRVIKAR\AppData\Local\Temp\ipykernel_14160\786792159.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['last_five'] = last_five


In [24]:
final_df = df.groupby('match_id').sum()['runs'].reset_index().merge(df,on='match_id')


In [25]:
final_df=final_df[['batting_team','bowling_team','city','current_score','balls_left','wickets_left','crr','last_five','runs_x']]


In [26]:
final_df.dropna(inplace=True)


In [27]:
final_df.isnull().sum()


batting_team     0
bowling_team     0
city             0
current_score    0
balls_left       0
wickets_left     0
crr              0
last_five        0
runs_x           0
dtype: int64

In [28]:
final_df = final_df.sample(final_df.shape[0])


In [29]:
final_df.sample(2)


batting_team bowling_team   city  current_score  balls_left  \
28194   Bangladesh     Pakistan  Dhaka             64          60   
24035   Bangladesh  New Zealand  Dhaka             52          66   

       wickets_left       crr  last_five  runs_x  
28194             7  6.400000       34.0     108  
24035            10  5.777778       31.0     141

In [30]:
X = final_df.drop(columns=['runs_x'])
y = final_df['runs_x']
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=1)

In [31]:
X_train

batting_team  bowling_team          city  current_score  balls_left  \
26528     Sri Lanka   West Indies     Abu Dhabi             38          94   
72936   New Zealand    Bangladesh        Mirpur             42          97   
73180   Afghanistan    Bangladesh        Mirpur             15         100   
20242   West Indies       England    Bridgetown             99          58   
45920  South Africa   West Indies       Tarouba             34          77   
...             ...           ...           ...            ...         ...   
3163       Pakistan   West Indies      Trinidad             45          79   
57428      Pakistan       England      Barbados             84          50   
18903         India       England     Ahmedabad            110          41   
47948   West Indies  South Africa  Johannesburg             10         112   
68336   West Indies   New Zealand    Lauderhill            136          20   

       wickets_left        crr  last_five  
26528            10   8.769231       38.0  
72936            10  10.956522       42.0  
73180             9   4.500000       15.0  
20242             8   9.580645       51.0  
45920             9   4.744186       29.0  
...             ...        ...        ...  
3163              8   6.585366       33.0  
57428             6   7.200000       33.0  
18903             7   8.354430       42.0  
47948            10   7.500000       10.0  
68336             6   8.160000       49.0  

[67864 rows x 8 columns]

In [32]:


# Column transformer for one-hot encoding
trf = ColumnTransformer([
    ('trf', OneHotEncoder(sparse_output=False, drop='first'), ['batting_team', 'bowling_team', 'city'])
], remainder='passthrough')

# List of models to compare
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Random Forest": RandomForestRegressor(n_estimators=100, random_state=1),
    "XGBoost": XGBRegressor(n_estimators=1000, learning_rate=0.2, max_depth=12, random_state=1),
    "LightGBM": LGBMRegressor(n_estimators=1000, learning_rate=0.2, max_depth=12, random_state=1)
}

# Dictionary to store results
results = {}

# Loop through each model, create a pipeline, fit, and evaluate
for name, model in models.items():
    pipe1 = Pipeline(steps=[
        ('step1', trf),
        ('step2', StandardScaler()),  # Scale the data
        ('step3', model)  # Model step
    ])
    
    # Fit the pipeline
    pipe1.fit(X_train, y_train)
    
    # Predictions
    y_pred = pipe1.predict(X_test)
    
    # Evaluate model performance
    r2 = r2_score(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)
    
    # Store results
    results[name] = {
        "R2 Score": r2,
        "Mean Absolute Error": mae
    }

# Print results for each model
for model_name, metrics in results.items():
    print(f"Model: {model_name}")
    print(f"R2 Score: {metrics['R2 Score']}")
    print(f"Mean Absolute Error: {metrics['Mean Absolute Error']}")
    print("-" * 30)


[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.010773 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 942
[LightGBM] [Info] Number of data points in the train set: 67864, number of used features: 75
[LightGBM] [Info] Start training from score 161.963825
Model: Logistic Regression
R2 Score: 0.8625262600299726
Mean Absolute Error: 3.6777862910355394
------------------------------
Model: Random Forest
R2 Score: 0.9445540607802703
Mean Absolute Error: 3.5926348032268054
------------------------------
Model: XGBoost
R2 Score: 0.9602288603782654
Mean Absolute Error: 2.7320116781321415
------------------------------
Model: LightGBM
R2 Score: 0.9507746973089413
Mean Absolute Error: 4.228131055962365
------------------------------


In [36]:
pipe = Pipeline(steps=[
    ('trf', trf),
    ('step2', StandardScaler()),  # Scale the data
    ('xgb_model', XGBRegressor(n_estimators=1000, learning_rate=0.2, max_depth=12, random_state=1))
])

# Fit the pipeline with your training data
pipe.fit(X_train, y_train)

# Save the pipeline
with open('finalpipe.pkl', 'wb') as f:
    pickle.dump(pipe, f)


In [34]:
pickle.dump(pipe,open('pipe.pkl','wb'))

The output you provided seems to be the result of training various machine learning models, including LightGBM, Logistic Regression, Random Forest, and XGBoost. Here’s a breakdown of the information presented, along with explanations of the key metrics and messages:

LightGBM Information:
Row-wise Multi-threading:

The LightGBM library is automatically choosing to use row-wise multi-threading for efficiency. The overhead time for this selection was approximately 0.0128 seconds.
You can set force_row_wise=true to eliminate this overhead if desired.
If you encounter memory issues, you can use force_col_wise=true to switch to column-wise processing, which can be more memory efficient but might be slower.
Total Bins:

Total Bins 942: This indicates that LightGBM has created 942 bins to represent the continuous features in the dataset, which is part of its optimization process.
Training Data Information:

Number of data points in the train set: 67864: This tells you that there are 67,864 samples (data points) used for training the model.
Number of used features: 75: This indicates that 75 features are being used in the training process.
Start Training Score:

Start training from score 162.011818: This is the initial score (baseline prediction) before the model starts training on the data.
Model Performance Metrics:
The output lists the performance metrics for various models, comparing them based on two metrics: R² Score and Mean Absolute Error (MAE).

Logistic Regression:

R² Score: 0.8624
Indicates that approximately 86.24% of the variance in the target variable is explained by the model.
Mean Absolute Error: 3.54
On average, the model's predictions deviate from the actual values by about 3.54 units.
Random Forest:

R² Score: 0.9415
This model explains about 94.15% of the variance.
Mean Absolute Error: 3.52
Average prediction error is slightly lower than that of Logistic Regression.
XGBoost:

R² Score: 0.9605
This indicates a high explanatory power, with 96.05% of the variance explained.
Mean Absolute Error: 2.64
XGBoost has the lowest average prediction error among the models listed.
LightGBM:

R² Score: 0.9506
The model explains 95.06% of the variance, showing strong performance.
Mean Absolute Error: 4.11
This model has the highest average prediction error of the four, which might indicate that it could benefit from tuning or additional feature engineering.
Summary:
Model Comparison: XGBoost performs the best in terms of both R² score and MAE, followed closely by Random Forest and LightGBM. Logistic Regression performs the least well among the four models.
Insights: The output suggests that tree-based models (XGBoost and Random Forest) generally perform better on the dataset than Logistic Regression, likely due to their ability to capture non-linear relationships in the data.
Model Tuning: Given the performance metrics, you may want to consider hyperparameter tuning for the LightGBM model to improve its accuracy or explore other model architectures if applicable.
This analysis provides valuable insights into model performance and helps in deciding which model to use for predictions based on the given dataset.